src/util.cc

```diff
@@ -604,3 +606,51 @@ bool Truncate(const string& path, size_t size, string* err) {
   }
   return true;
 }
+
+bool ReplaceContent(const string& path, const string& old_path, string* err) {
+#ifndef _WIN32
+  struct stat old_file;
+  bool found_uid_gid = true;
+
+  //get the stat
+  if (stat(path.c_str(), &old_file) < 0) {
+    //save if we found the stat
+    found_uid_gid = false;
+  }
+
+  if (unlink(path.c_str()) < 0) {
+    *err = strerror(errno);
+    return false;
+  }
+
+  if (rename(old_path.c_str(), path.c_str()) < 0) {
+    *err = strerror(errno);
+    return false;
+  }
+
+  //apply uid and gid again so we always stay as the uid and gid of the first caller that created a file
+  if (found_uid_gid) {
+    if (chown(path.c_str(), old_file.st_uid, old_file.st_gid)) {
+      *err = strerror(errno);
+      return false;
+    }
+  }
+#else
+  if (!CopyFile(old_path.c_str(), path.c_str(), false))
```

Comment:
This is the same behavior as the unlink/rename you're replacing, right? can we just keep using unlink/replace then?

And then that part could be shared still and it'd look like

```
#if !defined(_WIN32)
stat()
#endif
unlink()
rename()
#if !defined(_WIN32)
chown()
#endfi

Reply:
Well, as far as I remember CopyFile keeps the owner of the new file as long as it exists. And only copies the content, i cannot find any source anymore saying that. However, when doing only unlink() rename() the same issue we have on linux hits on windows ...

Reply:
Ah, thanks. Sounds subtle. I could imagine someone coming along and "simplifying" this code to what I just proposed some time later. I wonder if there's something we could do to stop them since using CopyFile() is important here. Do you think it's possible to write an automated test for that? Maybe create a file with non-standard permissions, call CopyFile() and check the new file has the non-standard permissions? Ideally in a way that the test fails if someone does the change I proposed here.